src/ray/core_worker/lib/golang/go_worker.cpp

```diff
@@ -0,0 +1,328 @@
+
+#include "ray/common/id.h"
+#include "ray/core_worker/core_worker.h"
+#include "ray/gcs/gcs_client/global_state_accessor.h"
+
+using namespace std;
+
+template <typename ID>
+inline ID ByteArrayToId(char *bytes) {
+  std::string id_str(ID::Size(), 0);
+  memcpy(&id_str.front(), bytes, ID::Size());
+  return ID::FromBinary(id_str);
+}
+
+DataBuffer *AllocateDataBuffer(void *ptr, int size) {
+  auto db = new DataBuffer();
+  db->p = ptr;
+  db->size = size;
+  return db;
+}
+
+RAY_EXPORT DataValue *go_worker_AllocateDataValue(void *data_ptr, size_t data_size,
+                                                  void *meta_ptr, size_t meta_size) {
+  auto dv = new DataValue();
+  dv->data = AllocateDataBuffer(data_ptr, data_size);
+  dv->meta = AllocateDataBuffer(meta_ptr, meta_size);
+  return dv;
+}
+
+RAY_EXPORT void go_worker_Initialize(int workerMode, char *store_socket,
+                                     char *raylet_socket, char *log_dir,
+                                     char *node_ip_address, int node_manager_port,
+                                     char *raylet_ip_address, char *driver_name,
+                                     int jobId, char *redis_address, int redis_port,
+                                     char *redis_password, char *serialized_job_config) {
+  auto task_execution_callback =
+      [](ray::TaskType task_type, const std::string task_name,
+         const ray::core::RayFunction &ray_function,
+         const std::unordered_map<std::string, double> &required_resources,
+         const std::vector<std::shared_ptr<ray::RayObject>> &args,
+         const std::vector<ObjectID> &arg_reference_ids,
+         const std::vector<ObjectID> &return_ids, const std::string &debugger_breakpoint,
+         std::vector<std::shared_ptr<ray::RayObject>> *results,
+         std::shared_ptr<ray::LocalMemoryBuffer> &creation_task_exception_pb) {
+        // convert RayFunction
+        auto function_descriptor = ray_function.GetFunctionDescriptor();
+        auto typed_descriptor = function_descriptor->As<ray::CppFunctionDescriptor>();
+
+        char *fd_data[1];
+        fd_data[0] = const_cast<char *>(typed_descriptor->FunctionName().c_str());
+        GoSlice fd_list;
+        fd_list.data = fd_data;
+        fd_list.len = 1;
+        fd_list.cap = 1;
+
+        std::vector<DataValue *> args_array_list;
+        for (auto &it : args) {
+          auto data = it->GetData();
+          auto meta = it->GetMetadata();
+          args_array_list.push_back(go_worker_AllocateDataValue(
+              data->Data(), data->Size(), meta->Data(), meta->Size()));
+        }
+
+        GoSlice args_go;
+        args_go.data = &args_array_list[0];
+        args_go.len = args_array_list.size();
+        args_go.cap = args_array_list.size();
+
+        std::vector<std::shared_ptr<DataValue>> return_value_list;
+        for (size_t i = 0; i < return_ids.size(); i++) {
+          return_value_list.push_back(make_shared<DataValue>());
```

Comment:
What is the purpose of making this `make_shared`? 

Since the return value list will get its pointers replaced by pointers generated from the callback without destructing the original ptr, this would result in memory leak.